tap_github/client.py

```diff
@@ -0,0 +1,271 @@
+        """
+        Fetch all pages of records and return them.
+        """
+        while True:
+            r = self.authed_get(source, url, headers)
+            yield r
+
+            # Fetch the next page if next found in the response.
+            if 'next' in r.links:
+                url = r.links['next']['url']
+            else:
+            # Break the loop if all pages are fetched.
+                break
+
+    def verify_repo_access(self, url_for_repo, repo):
+        """
+        Call rest API to verify that the user has sufficient permissions to access this repository.
+        """
+        try:
+            self.authed_get("verifying repository access", url_for_repo)
+            LOGGER.info("Verifying access of repository: %s", repo)
+        except NotFoundException:
+            # Throwing user-friendly error message as it checks token access
+            message = "HTTP-error-code: 404, Error: Please check the repository name \'{}\' or you do not have sufficient permissions to access this repository.".format(repo)
+            raise NotFoundException(message) from None
+
+    def verify_access_for_repo(self):
+        """
+        For all the repositories mentioned in the config, check the access for each repos.
+        """
+        access_token = self.config['access_token']
+        self.session.headers.update({'authorization': 'token ' + access_token, 'per_page': '1', 'page': '1'})
+
+        repositories = self.extract_repos_from_config()
+
+        for repo in repositories:
+
+            url_for_repo = "https://api.github.com/repos/{}/commits".format(repo)
+
+            # Verifying for Repo access
+            self.verify_repo_access(url_for_repo, repo)
+
+    def extract_repos_from_config(self):
+        """
+        Extracts all repositories from the config and calls get_all_repos()
+        for organizations using the wildcard 'org/*' format.
+        """
+        repo_paths = list(filter(None, self.config['repository'].split(' ')))
+
+        orgs_with_all_repos = list(filter(lambda x: x.split('/')[1] == '*', repo_paths))
+
+        if orgs_with_all_repos:
+            # Remove any wildcard "org/*" occurrences from `repo_paths`
+            repo_paths = list(set(repo_paths).difference(set(orgs_with_all_repos)))
+
+            # Get all repositories for an org in the config
+            all_repos = self.get_all_repos(orgs_with_all_repos)
+
+            # Update repo_paths
+            repo_paths.extend(all_repos)
+
+        return repo_paths
+
+    def get_all_repos(self, organizations: list):
+        """
+        Retrieves all repositories for the provided organizations and
+        verifies basic access for them.
+
+        Docs: https://docs.github.com/en/rest/reference/repos#list-organization-repositories
+        """
+        repos = []
+
+        for org_path in organizations:
+            org = org_path.split('/')[0]
+            for response in self.authed_get_all_pages(
```

Comment:
I have checked that tap will throw `NotFoundException` in case of error code 404.

Reply:
I don't think so, it will throw NotfoundException during sync mode execution. The code flow does not include this exception anywhere. Can you please redirect to the line from where this exception will thrown?